### ISSUE TITLE
Electron SUID sandbox

### ISSUE THREAD

**Issue by fxha** [flatpak/flatpak#3044]
## Linux distribution and version

Fedora 30

## Flatpak version

`1.4.2`

## Description of the problem

Electron `>=v5` will fail to execute due no kernel access and invalid sandbox [binary permissions](https://github.com/electron/electron/issues/17972). Chromiums sandbox need user namespaces or the sandbox binary (`chrome-sandbox`) must be installed with SUID flag and root owner. Is there a way to configure the package that we can execute Electron with working sandbox? Currently `--no-sandbox` flag is needed to get a Electron v4 no renderer process sandbox behavior. Is it possible to own a file as root in the flatpak sandbox?

```
[2:0805/160109.913391:FATAL:setuid_sandbox_host.cc(157)] The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /app/<name>/chrome-sandbox is owned by root and has mode 4755.
```

## Steps to reproduce

Build Electron `>=v5` as flatpak and try to execute it.

**Comment by refi64** [comment]
Not possible, I guess the urgency of my previous Chromium-related experiments just shot up...

**Comment by gasinvein** [comment]
Isn't it more like a bubblewrap issue than flatpak's?

**Comment by Erick555** [comment]
No, `chrome-sandbox` needs suid bit (or unpriv user namespaces) and flatpak doesn't allow installing suid binaries (also it mounts fs with `nosuid`). This is the reason chrome isn't on flathub.` --no-sandbox` seems the only option.

**Comment by SISheogorath** [comment]
https://github.com/electron/electron/issues/17972#issuecomment-516957971 <-- at least according to this comment we might should raise and issue in the electron repository

**Comment by fxha** [comment]
@SISheogorath I think this is not a Chromium/Electron issue because Flatpak (or the sandbox Flatpak is using) don't provide a way to use Chromium's security features. From Chromium's perspective, it's the correct way to hard fail because the needed requirements are not meet. The issue with Chromium is that Flatpak's sandbox isolate the application and we as package maintainers cannot set the necessary permissions on the "fallback" sandbox binary. E.g. when installing `deb` or `rpm` packages, you can set the necessary permissions and the binary is automatically owned by `root`. So you have a fallback if user namspaces are not supported by the distro.

**Comment by andrunko** [comment]
Related: https://discourse.gnome.org/t/sandboxing-portal/1651.

**Comment by enricotagliavini** [comment]
SUID is just a fallback in case clone() with NEWUSER flag fails. Running a modern unpatched kernel requires no privileges to do that. However it seems tricky to do from inside an already made sandbox.

Would it be possible to (securely) allow something similar to [https://stackoverflow.com/questions/46140088/why-is-unprivileged-recursive-unshareclone-newuser-not-permitted](https://stackoverflow.com/questions/46140088/why-is-unprivileged-recursive-unshareclone-newuser-not-permitted)

However uid and gid maps seems to be set correctly in the sandbox as far as I understand, but then I don't get why the clone() syscall fails with EPERM 

```
19719 readlink("/proc/self/exe", "/app/Signal/signal-desktop", 4096) = 26
19719 access("/app/Signal/chrome-sandbox", F_OK) = 0
19719 getresuid([1000], [1000], [1000]) = 0
19719 getresgid([1000], [1000], [1000]) = 0
19719 clone(child_stack=0x7fffbcdc6410, flags=CLONE_NEWUSER|SIGCHLD) = -1 EPERM (Operation not permitted)
19719 lstat("/app", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
19719 lstat("/app/Signal", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
19719 lstat("/app/Signal/signal-desktop", {st_mode=S_IFREG|0755, st_size=112456472, ...}) = 0
19719 socketpair(AF_UNIX, SOCK_SEQPACKET, 0, [11, 12]) = 0
19719 setsockopt(11, SOL_SOCKET, SO_PASSCRED, [1], 4) = 0
19719 access("/app/Signal/chrome-sandbox", F_OK) = 0
19719 stat("/app/Signal/chrome-sandbox", {st_mode=S_IFREG|S_ISUID|0755, st_size=235840, ...}) = 0
19719 access("/app/Signal/chrome-sandbox", X_OK) = 0
19719 getpid()                          = 2
19719 write(2, "[2:0907/113237.256171:FATAL:setuid_sandbox_host.cc(157)] The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /app/Signal/chrome-sandbox is owned by root and has mode 4755.\n", 275) = 275
```

The only other documented EPERM related to NEWUSER is

> CLONE_NEWUSER was specified in flags and the caller is in a chroot environment (i.e., the caller's root directory does not match the root directory of the mount namespace in which it resides).

but I haven't got a clue how to check that.

**Comment by Erick555** [comment]
> Running a modern unpatched kernel requires no privileges to do that.

Every kernel in Debian requires privileges for that. Maybe some other distros too

**Comment by enricotagliavini** [comment]
> > Running a modern unpatched kernel requires no privileges to do that.
> 
> Every kernel in Debian requires privileges for that. Maybe some other distros too

Very true, but that's a deliberate Debian decision of not following upstream. I understand their reasons, and they have all the rights to do so.

However when running upstream-like kernels where unprivileged user namespace is available it should be used. A generic check for application requiring such a feature can be implemented and issue a warning they are being started with --no-sandbox when running on Debian or kernel without said feature.

If the above problem has a solution at all :)

**Comment by Erick555** [comment]
> A generic check for application requiring such a feature can be implemented and issue a warning they are being started with --no-sandbox when running on Debian or kernel without said feature.

For debian like kernels they have setuid sandbox. `--no-sandbox` may be acceptable for flatpak which has its own sandboxing but for distros like Debian it will be highly questionable. Also there were reports that `--no-sandbox` didn't work for skype at all.

**Comment by Esokrates** [comment]
IMO you should look for a way to make the chromium sandbox work, compromising upstream security is not what any downstream / distributor should do. You do way more harm disabling the chromium sandbox than allowing suid binaries in Flatpak.

**Comment by TingPing** [comment]
I believe we'd all love that and it is the end goal. However nobody here is a Chromium contributor and probably doesn't even have a machine that can compile that project.

**Comment by refi64** [comment]
Some background: I've made patches to implement this before, but from what
I can tell, upstream has no interest in them. They basically said they'd be
interested if Flatpak became the main Linux packaging standard, but as of
right now, it's not. Afaik this was part the background for @ramcq's
interest in making a cross-packaging sandbox API: it would be a much easier
sell if we could say "any packaging / sandbox tool can implement this, and
one (or two) major players already do.
*.

I've also used LD_PRELOAD to fake having an suid sandbox before, which
works-ish, but right now I've got to basically make a custom Zygote of
sorts for it that...isn't actually a Zygote, so that's going to take a
bit of time to be ready for actual usage. (It would be great if --no-zygote
could just be used, but using that without --no-sandbox causes assertion
failures that can't be worked around.)

The TLDR would be, "it's probably not getting upstreamed soon, and a
workaround exists, but it's not ready yet".

**Comment by mat8913** [comment]
From what I can tell, namespaces in flatpak are explicitly disallowed by https://github.com/flatpak/flatpak/blob/1f35dda1b50cceb24c471b19ac886dc91bf8a0b4/common/flatpak-run.c#L2481-L2485

Would it be possible to add an option (`--allow=namespaces` perhaps) to allow it? Or would an app be able to use that to break free of flatpak's sandbox?

**Comment by TingPing** [comment]
It is blacklisted because it is considred a security risk.

**Comment by ramcq** [comment]
Alex is planning to work on a patch for Chromium/Electron which will ask Flatpak to set up the sandbox using the existing (or similar, if new capabilities are needed) API. The hope is that we can identify an API which can be provided as a portal so that both Snap and Flatpak can implement it. https://discourse.gnome.org/t/sandboxing-portal/1651

**Comment by refi64** [comment]
While everything else is taking place, there's currently [zypak](http://github.com/refi64/zypak) which can be used for Electron apps to make them use the native Flatpak sandboxing primitives.

**Comment by SISheogorath** [comment]
@refi64 are you currently implementing it in the electron base flatpak? Would look forward to get this running for Riot where the PR was closed a while ago.

**Comment by malept** [comment]
Not @refi64, but I'm trying to get it working in [`@malept/electron-installer-flatpak`](https://github.com/malept/electron-installer-flatpak), which is obviously on a different level from the base flatpak.

**Comment by Erick555** [comment]
The zypak solution exist and works well so I think this issue should be closed as app maintainers are being misled that this is still unsolved problem and choose the wrong workaround.

cc @TingPing @barthalion

**Comment by soredake** [comment]
There is problem with electron games in flatpak, they will not work out of the box, requiring you to set `--no-sandbox` manually, or try to convince maintainer to use zypak/set --no-sandbox when flatpak is detected, which is not going to work.
